bricks/strings/regex.h

```diff
@@ -0,0 +1,192 @@
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*******************************************************************************/
+
+#ifndef BRICKS_STRINGS_REGEX_H
+#define BRICKS_STRINGS_REGEX_H
+
+#include <regex>
+#include <string>
+#include <unordered_map>
+
+namespace current {
+namespace strings {
+
+// A simple way to range-based-for-loop iterate over regex matches.
+struct IterateByRegexMatches {
+  const std::sregex_iterator begin_;
+  const std::sregex_iterator end_ = std::sregex_iterator();
+
+  template <typename T>
+  IterateByRegexMatches(const std::regex& re, T&& s)
+      : begin_(s.begin(), s.end(), re) {}
+
+  std::sregex_iterator begin() const { return begin_; }
+  std::sregex_iterator end() const { return end_; }
+};
+
+// Regexes in C++ don't support named capture. In Current, we have a workaround. Refresher:
+// 1) `(blah)` is an unnamed capture group (referencable to by its index in the regex though).
+// 2) `(?:blah)` is a non-capturing group (should match, but can't be referred to).
+// 3) `(?<whoa>blah)` is a named capture -- the one C++ does not support -- allowing referring to "blah" as `whoa`.
+// 4) Obviously, `\(` is an opening parenthesis, not the beginning of the group.
+class NamedRegexCapturer {
+ private:
+  struct Data {
+    const std::vector<std::string> group_names;
+    const std::unordered_map<std::string, size_t> group_indexes;
+    const std::regex transformed_re;
+    static Data Construct(const std::string& re_body) {
+      std::vector<std::string> names;
+      std::unordered_map<std::string, size_t> indexes;
+      // This regex requires explanation.
+      // 1) Must begin with a non-escaped "(".
+      // 2) Must not begin with "(?:".
+      // 3) May, but does not have to be "(?<...>").
+      const std::regex re_capture_groups("\\((?!\\?:)(\\?<(\\w+)>)?");
+      for (const auto& capture_group : current::strings::IterateByRegexMatches(re_capture_groups, re_body)) {
+        if (!capture_group.position() || re_body[capture_group.position() - 1u] != '\\') {
+          const std::string name = capture_group[2].str();
+          if (!name.empty()) {
+            indexes[name] = names.size() + 1u;
+          }
+          names.push_back(name);
+        }
+      }
+      const std::string cleaned_re_body = std::regex_replace(re_body, re_capture_groups, "(");
+      return Data{names, indexes, std::regex(cleaned_re_body)};
```

Comment:
`std::move()` for `names` and `indexes`?

Reply:
:+1: 80c33dd4